### ISSUE TITLE
Request: @namedResult

### ISSUE THREAD

**Issue by mcandre** [go-critic/go-critic#1480]
Hi,

I would like a rule that behaves essentially opposite to `@unnamedResult`. Instead of encouraging programmers to name individual tuple field members in return signatures, I want a warning when Go code uses named returns.

Prior art:

* golantci-lint (nakedret, nonamedreturns) dead
* lint-naked-returns dead

I like the overall design of go critic. So far, the rules enabled out of the box appear fine.

I humbly request that `@namedResult` be enabled by default. Since the other one, `@unnamedResult` is already disabled by default, then there should be no conflict out of the box. Some fanatics insist on using named returns, in which case they can configure go critic accordingly. If backwards compatibility is at a premium, then we may leave `@namedResult` disabled by default.

**Comment by quasilyte** [comment]
Should it just report all `return` statements without values of non-void functions?

Seems doable in ruleguard syntax if that's it. Or maybe I'm missing some corner cases.

**Comment by cristaloleg** [comment]
> I want a warning when Go code uses named returns.

What is wrong with `func f() (x, y float64)` ?

**Comment by quasilyte** [comment]
AFAIK, this issue is just about the nakedret replacement. But @mcandre needs to confirm that

**Comment by mcandre** [comment]
> What is wrong with func f() (x, y float64) ?

That line is fine, but there are some contexts, and many Go programmers, who wish to prohibit named returns for certain needs.

For example, named returns tend to worsen any variable shadowing conflicts (as an aside, would be good for go-critic to automatically warn on shadowing, like go vet but enabled by default). Error handling becomes a nightmare when people accidentally shadow `err`. For my part, I use a notation of incrementing a binding suffix, `err2`, `err3`, etc. in each successive lexical scope nesting. But not everyone applies such safety measures.

I admit that for simple tuples, named returns may offer a productive balance between clarity on the one hand and structless simplicity on the other. More complicated returns should adopt a struct-like mechanism.

One rule of thumb for maintainability is to promote arguments or returns to their own structs, is when the number of values reaches four, though of course that's an arbitrary distinction.

However, other Go uses of named returns are not so good as numerical equations.

A single return value is unambiguous, so no need to clarify the semantic order of values.

In general, I like the option of warning on named returns (all contexts). I respect that other programmers may want this rule disabled, such as with per-line linter configuration comments, CLI flags, or even disabled by default. I won't push aggressively to enable the rule by default, but I do want the option to exist.

**Comment by cristaloleg** [comment]
> The main reason is some people, me included, enable all go-critric rules via golangci-lint

Enabling all checks without taking a look on them is a wrong strategy (sad but true).

> so they would have it reported twice

Sounds like a simple fix to lint config, isn't it? I mean it's not like make a patch to 3rd party tool or so, just a few lines in YAML.

> do not have multiple linter for the same thing

`golangci-lint` is a great project but `go-critic` is/can be used independently. Also, there are duplicates already (revive, staticcheck, go-critic vs one-pattern linters)

**Comment by ccoVeille** [comment]
Yes, everything you said is true.

**Comment by mcandre** [comment]
Nah, that's a good point about the existing rule conflicting with a more flexible rule. I also enable all possible checks. Time is too valuable individually opting into every single rule.

Yes, enabling all rules in a linter is expected to occasionally change behavior in a backwards incompatible, breaking way. If that's too much noise, setup a GitHub Actions job and decouple it from the ruleset. So that you get notified of warnings / breakages early, without impacting pull request controls.

**Comment by quasilyte** [comment]
> So here, what happens when people have both of them enabled. 

Nakedret is enabled for many people already via golangci-lint. This is not about prohibiting named results, but about prohibiting naked return statement. At least that's my understanding (confirmed by @mcandre to be aligned with the issue)

```go
// Good
func f() (x, y int) {
  return x, y
}

// Bad
func f() (x, y int) {
  return
}
```

**Comment by mcandre** [comment]
Indeed, nakedret does this. Until very recently, it was broken on the latest Go releases. I still have to check to ensure that it works with Go 1.24.2.

**Comment by cristaloleg** [comment]
Also

> golantci-lint (nakedret, nonamedreturns) dead

Any proofs? https://github.com/alexkohler/nakedret & https://github.com/firefart/nonamedreturns have recent commits. Can you specify what is wrong with them?

**Comment by danrneal** [comment]
My preference would be for a complementary checker to unnamedResult. Something like unnamedResult flags whenever a result should be named (such as when the results are both of the same type). And then the new checker to flag all other times where a result is named but would not have been flagged by unnamedResult. And then likewise banning naked returns entirely.
